File: src/Core/FinancialEvents/Schemas/Grouping.py
from dataclasses import dataclass, field
from typing import Sequence

import Core.FinancialEvents.Schemas.CommonFormats as cf
from Core.FinancialEvents.Schemas.Events import (
    TradeEventDerivativeAcquired,
    TradeEventDerivativeSold,
    TradeEventStockAcquired,
    TradeEventStockSold,
    TransactionCash,
)
from Core.FinancialEvents.Schemas.FinancialIdentifier import FinancialIdentifier
from Core.FinancialEvents.Schemas.Provenance import ProvenanceStep
from Core.FinancialEvents.Schemas.Lots import TaxLotDerivative, TaxLotStock


@dataclass
class DerivativeGrouping:
    FinancialIdentifier: FinancialIdentifier
    DerivativeTrades: Sequence[TradeEventDerivativeAcquired | TradeEventDerivativeSold]
    DerivativeTaxLots: Sequence[TaxLotDerivative]
    Provenance: Sequence[ProvenanceStep] = field(default_factory=list)


@dataclass
class FinancialGrouping:
    FinancialIdentifier: FinancialIdentifier
    CountryOfOrigin: str | None  # None for unknown

    UnderlyingCategory: cf.GenericCategory

    StockTrades: Sequence[TradeEventStockAcquired | TradeEventStockSold]
    StockTaxLots: Sequence[TaxLotStock]

    DerivativeGroupings: Sequence[DerivativeGrouping]

    CashTransactions: Sequence[TransactionCash]
    Provenance: Sequence[ProvenanceStep] = field(default_factory=list)


@dataclass
class UnderlyingDerivativeGrouping:
    FinancialIdentifier: FinancialIdentifier
    DerivativeTrades: Sequence[TradeEventDerivativeAcquired | TradeEventDerivativeSold]


@dataclass
class UnderlyingGroupingWithTradesOfInterest:
    FinancialIdentifier: FinancialIdentifier
    CountryOfOrigin: str | None  # None for unknown

    UnderlyingCategory: cf.GenericCategory

    StockTrades: Sequence[TradeEventStockAcquired | TradeEventStockSold]

    DerivativeGroupings: Sequence[UnderlyingDerivativeGrouping]

    CashTransactions: Sequence[TransactionCash]

File: lib/core/schemas/Grouping.ts
import type { GenericCategory } from "./CommonFormats.ts";
import type { FinancialIdentifier } from "./FinancialIdentifier.ts";
import type { AnyProvenanceStep } from "./Provenance.ts";
import type {
	TradeEventDerivativeAcquired,
	TradeEventDerivativeSold,
	TradeEventStockAcquired,
	TradeEventStockSold,
	TransactionCash,
} from "./Events.ts";
import type { TaxLotDerivative, TaxLotStock } from "./Lots.ts";

export type DerivativeGrouping = {
	financialIdentifier: FinancialIdentifier;
	derivativeTrades: (TradeEventDerivativeAcquired | TradeEventDerivativeSold)[];
	derivativeTaxLots: TaxLotDerivative[];
	provenance: AnyProvenanceStep[];
};

export type FinancialGrouping = {
	financialIdentifier: FinancialIdentifier;
	countryOfOrigin: string | null;

	underlyingCategory: GenericCategory;

	stockTrades: (TradeEventStockAcquired | TradeEventStockSold)[];
	stockTaxLots: TaxLotStock[];

	derivativeGroupings: DerivativeGrouping[];

	cashTransactions: TransactionCash[];
	provenance: AnyProvenanceStep[];
};

export type UnderlyingDerivativeGrouping = {
	financialIdentifier: FinancialIdentifier;
	derivativeTrades: (TradeEventDerivativeAcquired | TradeEventDerivativeSold)[];
};

export type UnderlyingGroupingWithTradesOfInterest = {
	financialIdentifier: FinancialIdentifier;
	countryOfOrigin: string | null;

	underlyingCategory: GenericCategory;

	stockTrades: (TradeEventStockAcquired | TradeEventStockSold)[];

	derivativeGroupings: UnderlyingDerivativeGrouping[];

	cashTransactions: TransactionCash[];
};
